File: deep500/frameworks/reference/custom_operators/cpp/operators/batchnormalization_op_training.cpp
#include <deque>
#include "deep500/deep500.h"
#include "../utility/tensor/copy_data.hpp"
#include "../utility/tensor/set_all.hpp"
#include "../utility/math/mean.hpp"
#include "../utility/math/variance.hpp"
#include <iostream>

//batchnormalization - training mode

template<typename T>
class BatchNormalizationLayerTraining : public deep500::CustomOperator{

protected:
    const float m_epsilon;
    const float m_momentum;
    const int m_spatial;
    int m_initalized; //running mean and running var are init. in first call of forward
    const std::deque<uint32_t> m_input_shape;
    uint32_t m_input_size;
    
public: 
    //constructor
    BatchNormalizationLayerTraining(
        const std::deque<uint32_t> input_shape,
        const float epsilon,
        const float momentum,
        const int spatial) :
        m_epsilon(epsilon),
        m_momentum(momentum),
        m_spatial(spatial),
        m_input_shape(input_shape) {

        //compute numpber of elements in input
        m_input_size = 1;
        for(const auto& elem: input_shape)
            m_input_size *= elem;
    }

    //copy constructor

    //destructor
    virtual ~BatchNormalizationLayerTraining() {}

    //member functions
    virtual bool supports_cuda() { return false; }

    void forward(
        const T *input_tensor,
        const T *input_tensor2,
        const T *input_tensor3,
        const T *input_tensor4,
        const T *input_tensor5,
        T *output_tensor,
        T *output_tensor2,
        T *output_tensor3,
        T *output_tensor4,
        T *output_tensor5) {
        
        /*
        ONNX notation:
        X = input1
        scale = input2
        B = input3
        mean = input4
        var = input5
        Y = output
        */

        //compute mean of minibatch
        utilityMath::compute_mean(m_input_shape, input_tensor, output_tensor4);

        //compute variance of minibatch
        utilityMath::compute_variance(m_input_shape, input_tensor, output_tensor4, output_tensor5);

        for(size_t i = 0; i < m_input_shape[1]; i++) {
            //compute running mean
            output_tensor2[i] = input_tensor2[i] * m_momentum + input_tensor4[i] * (1 - m_momentum);

            //compute running var
            output_tensor3[i] = input_tensor3[i] * m_momentum + input_tensor5[i] * (1 - m_momentum);
        }

        size_t num_blocks;
        size_t length_block;
        size_t num_elements;
        size_t spacing;
        size_t initial_offset;
        size_t temp;

        num_blocks = m_input_shape[0];

        temp = 1;
        for(size_t j = 2; j < m_input_shape.size(); j++) {
            temp *= m_input_shape[j];
        }
        length_block = temp;

        num_elements = num_blocks * length_block;
        spacing = m_input_shape[1] * length_block; 

        //compute standard deviation of minibatch
        //m_input_shape[1] = size of saved_mean
        T sqrtvar[m_input_shape[1]];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            sqrtvar[i] = std::sqrt(output_tensor5[i] + m_epsilon);
        }

        //compute output
        //assume: T output_tensor[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    output_tensor[initial_offset + k * spacing + l] = 
                    input_tensor2[i] * (input_tensor[initial_offset + k * spacing + l] -
                    output_tensor4[i]) / sqrtvar[i] + input_tensor3[i];
                }
            }
        }

        return;
    }

    void backward(
        const T *nextop_grad,
        const T *nextop_grad2,
        const T *nextop_grad3,
        const T *nextop_grad4,
        const T *nextop_grad5,
        const T *fwd_input_tensor,
        const T *fwd_input_tensor2,
        const T *fwd_input_tensor3,
        const T *fwd_input_tensor4,
        const T *fwd_input_tensor5,
        const T *fwd_output_tensor,
        const T *fwd_output_tensor2,
        const T *fwd_output_tensor3,
        const T *fwd_output_tensor4,
        const T *fwd_output_tensor5,
        T *input_tensor_grad,
        T *input_tensor2_grad,
        T *input_tensor3_grad,
        T *input_tensor4_grad,
        T *input_tensor5_grad) {

        utilityTensor::set_all_zero(m_input_size, input_tensor_grad);
        utilityTensor::set_all_zero(m_input_shape[1], input_tensor2_grad);
        utilityTensor::set_all_zero(m_input_shape[1], input_tensor3_grad);
        utilityTensor::set_all_zero(m_input_shape[1], input_tensor4_grad);
        utilityTensor::set_all_zero(m_input_shape[1], input_tensor5_grad);


        size_t num_blocks;
        size_t length_block;
        size_t num_elements;
        size_t spacing;
        size_t initial_offset;
        size_t temp;

        num_blocks = m_input_shape[0];

        temp = 1;
        for(size_t j = 2; j < m_input_shape.size(); j++) {
            temp *= m_input_shape[j];
        }
        length_block = temp;

        //number of elements in minibatch per channel
        num_elements = num_blocks * length_block;
        spacing = m_input_shape[1] * length_block;

        T x_min_mu[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    x_min_mu[initial_offset + k * spacing + l] = 
                    fwd_input_tensor[initial_offset + k * spacing + l] - fwd_output_tensor4[i];
                }
            }
        }

        T sqrtvar[m_input_shape[1]];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            sqrtvar[i] =  std::sqrt(fwd_output_tensor5[i] + m_epsilon);
        }
        
        T ivar[m_input_shape[1]];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            ivar[i] =  1. / sqrtvar[i];        
        }

        T x_hat[m_input_size];
        T d_x_hat[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    x_hat[initial_offset + k * spacing + l] = x_min_mu[initial_offset + k * spacing + l] * ivar[i];
                    d_x_hat[initial_offset + k * spacing + l] = fwd_input_tensor2[i] * nextop_grad[initial_offset + k * spacing + l];
                }
            }
        }
        
        T d_w[m_input_shape[1]];
        T d_b[m_input_shape[1]];
        utilityTensor::set_all_zero(m_input_shape[1], d_w);
        utilityTensor::set_all_zero(m_input_shape[1], d_b);
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_w[i] += nextop_grad[initial_offset + k * spacing + l] * x_hat[initial_offset + k * spacing + l];
                    d_b[i] += nextop_grad[initial_offset + k * spacing + l]; 
                }
            }
            
            
        }

        T d_ivar[m_input_shape[1]];
        utilityTensor::set_all_zero(m_input_shape[1], d_ivar);
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_ivar[i] += d_x_hat[initial_offset + k * spacing + l] * x_min_mu[initial_offset + k * spacing + l];
                }
            }
        }

        T d_xmu1[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_xmu1[initial_offset + k * spacing + l] = d_x_hat[initial_offset + k * spacing + l] * ivar[i];
                }
            }
        }

        T d_sqrtvar[m_input_shape[1]];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            d_sqrtvar[i] = -1. / std::pow(sqrtvar[i], 2) * d_ivar[i];
            
            
        }
        
        T d_var[m_input_shape[1]];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            d_var[i] = 0.5 * 1. / sqrtvar[i] * d_sqrtvar[i];
            
            
        }

        T d_sq[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_sq[initial_offset + k * spacing + l] = 1. / (T) num_elements * d_var[i];
                }
            }
        }

        T d_xmu2[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_xmu2[initial_offset + k * spacing + l] = 2 * x_min_mu[initial_offset + k * spacing + l] * d_sq[initial_offset + k * spacing + l];
                }
            }
        }

        T d_x1[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_x1[initial_offset + k * spacing + l] = d_xmu1[initial_offset + k * spacing + l] + d_xmu2[initial_offset + k * spacing + l];
                }
            }
        }

        T d_mu[m_input_shape[1]];
        utilityTensor::set_all_zero(m_input_shape[1], d_mu);
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_mu[i] += (-1.) * (d_xmu1[initial_offset + k * spacing + l] + d_xmu2[initial_offset + k * spacing + l]);
                }
            }
        }

        T d_x2[m_input_size];
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    d_x2[initial_offset + k * spacing + l] = 1. / (T) num_elements * d_mu[i];
                }
            }
        }

        //bwd_params
        for(size_t i = 0; i < m_input_shape[1]; i++) {
            initial_offset = i * length_block;
            
            for(size_t k = 0; k < num_blocks; k++) {
                for(size_t l = 0; l < length_block; l++) {
                    input_tensor2_grad[i] += nextop_grad[initial_offset + k * spacing + l] * (fwd_input_tensor[
                        initial_offset + k * spacing + l] - fwd_output_tensor4[i]) / sqrtvar[i];
                    input_tensor3_grad[i] += nextop_grad[initial_offset + k * spacing + l];
                }
            }
        }

        for(size_t i = 0; i < m_input_size; i++) {
            input_tensor_grad[i] = d_x1[i] + d_x2[i];
        }

        utilityTensor::set_all_zero(m_input_shape[1], input_tensor4_grad);
        utilityTensor::set_all_zero(m_input_shape[1], input_tensor5_grad);

        return;
    }


};

D500_EXPORTED void *create_new_op(deep500::tensor_t *input_descriptors, 
                                    int num_inputs,
                                    deep500::tensor_t *param_descriptors,
                                    int num_params,
                                    deep500::tensor_t *output_descriptors,
                                    int num_outputs) {

    std::deque<uint32_t> input_tensor_shape(
        input_descriptors[0].sizes,
        input_descriptors[0].sizes + input_descriptors[0].dims
    );

    return new BatchNormalizationLayerTraining<DTYPE>(input_tensor_shape, EPSILON, MOMENTUM, SPATIAL);
}

D500_REGISTER_OP(BatchNormalizationLayerTraining<DTYPE>);

File: deep500/frameworks/reference/custom_operators/python/batchnormalization_op_training.py
import numpy as np
from deep500.lv0.operators.operator_interface import CustomPythonOp

class BatchNormalizationTrainingOp(CustomPythonOp):
    def __init__(self, input_descriptors, output_descriptors, epsilon=1e-05, momentum=0.9, spatial=1, is_test=False):
        super(BatchNormalizationTrainingOp, self).__init__(input_descriptors, output_descriptors)
        self._input_desc = input_descriptors
        self._output_desc = output_descriptors
        
        self.epsilon = epsilon
        self.momentum = momentum
        self.spatial = spatial

    def forward(self, X, scale, B, mean, var):

        #N = number of elements in minibatch per channel
        N = X.shape[0]
        for i in range(2, len(X.shape)):
            N = N * X.shape[i]

        #compute mean
        saved_mean = np.copy(X)
        for i in range(2, len(X.shape)):
            saved_mean = np.sum(saved_mean, axis=i, keepdims=True)
        saved_mean = np.sum(saved_mean, axis=0)
        saved_mean = (1. / N) * saved_mean

        #compute variance
        xmu = X - saved_mean
        sq = xmu ** 2
        saved_var= np.copy(sq)
        for i in range(2, len(sq.shape)):
            saved_var = np.sum(saved_var, axis=i, keepdims=True)
        saved_var = np.sum(saved_var, axis=0)
        saved_var = (1. / N) * saved_var
         
        #compute running mean
        saved_mean = np.squeeze(saved_mean)
        running_mean = mean * self.momentum + saved_mean * (1 - self.momentum)

        #compute running variance
        saved_var = np.squeeze(saved_var)
        running_var = var * self.momentum + saved_var * (1 - self.momentum) 

        dims_x = len(X.shape)
        dim_ones = (1,) * (dims_x - 2)
        scale = scale.reshape(-1, *dim_ones)
        B = B.reshape(-1, *dim_ones)
        saved_mean = saved_mean.reshape(-1, *dim_ones)
        saved_var = saved_var.reshape(-1, *dim_ones)

        out = scale * (X - saved_mean) / np.sqrt(saved_var + self.epsilon) + B

        return [
            out,
            running_mean,
            running_var,
            np.reshape(saved_mean, running_mean.shape),
            np.reshape(saved_var, running_var.shape)]

    def backward(self, grads, fwd_inputs, fwd_outputs):
        #similar to source: https://kratzert.github.io/2016/02/12/understanding-the-gradient-flow-through-the-batch-normalization-layer.html

        x = fwd_inputs[0]
        saved_mean = fwd_outputs[3]
        saved_var = fwd_outputs[4]
        
        dims_x = len(x.shape)
        dim_ones = (1,) * (dims_x - 2)

        #N = number of elements in minibatch per channel
        N = x.shape[0]
        for i in range(2, len(x.shape)):
            N = N * x.shape[i]

        saved_mean = saved_mean.reshape(-1, *dim_ones)

        xmu = x - saved_mean

        saved_var = saved_var.reshape(-1, *dim_ones)

        sqrtvar = np.sqrt(saved_var + self.epsilon)
        
        ivar = 1./sqrtvar

        xhat = xmu * ivar

        dbeta = np.copy(grads[0])
        for i in range(2, len(grads[0].shape)):
            dbeta = np.sum(dbeta, axis=i, keepdims=True)
        dbeta = np.sum(dbeta, axis=0, keepdims=True)
        
        dgammax = grads[0]
        dgamma = np.copy(dgammax*xhat)
        for i in range(2, len((dgammax*xhat).shape)):
            dgamma = np.sum(dgamma, axis=i, keepdims=True)
        dgamma = np.sum(dgamma, axis=0, keepdims=True)
        
        scale = np.copy(fwd_inputs[1])
        scale = scale.reshape(-1, *dim_ones)
        
        dxhat = dgammax * scale

        divar = np.copy(dxhat*xmu)
        for i in range(2, len((dxhat*xmu).shape)):
            divar = np.sum(divar, axis=i, keepdims=True)
        divar = np.sum(divar, axis=0)
        
        dxmu1 = dxhat * ivar

        dsqrtvar = -1. /(sqrtvar**2) * divar

        dvar = 0.5 * 1. /np.sqrt(saved_var+self.epsilon) * dsqrtvar

        dsq = 1. /N * np.ones(fwd_inputs[0].shape) * dvar

        dxmu2 = 2 * xmu * dsq

        dx1 = (dxmu1 + dxmu2)
        
        dmu = np.copy(dxmu1 + dxmu2)
        for i in range(2, len((dxmu1 + dxmu2).shape)):
            dmu = np.sum(dmu, axis=i, keepdims=True)
        dmu = np.sum(dmu, axis=0)
        dmu = -1 * dmu

        dx2 = 1. /N * np.ones(fwd_inputs[0].shape) * dmu

        dx = dx1 + dx2

        return [
            dx.reshape(fwd_inputs[0].shape), 
            dgamma.reshape(fwd_inputs[1].shape),
            dbeta.reshape(fwd_inputs[2].shape),
            np.zeros_like(fwd_inputs[3]),
            np.zeros_like(fwd_inputs[4])]